packages/main/src/methods/chat-session-helpers.ts

```diff
@@ -0,0 +1,92 @@
+/**
+ * @license
+ * Copyright 2023 Google LLC
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import { Content, POSSIBLE_ROLES, Part, Role } from "../../types";
+import { GoogleGenerativeAIError } from "../errors";
+
+const VALID_PART_FIELDS: Array<keyof Part> = [
+  "text",
+  "inlineData",
+  "functionCall",
+  "functionResponse",
+];
+
+const VALID_PARTS_PER_ROLE: { [key in Role]: Array<keyof Part> } = {
+  user: ["text", "inlineData"],
+  function: ["functionResponse"],
+  model: ["text", "functionCall"],
+};
+
+const VALID_PREVIOUS_CONTENT_ROLES: { [key in Role]: Role[] } = {
```

Comment:
Is this documented anywhere or do we just know this by trial and error?

Reply:
Partially here: https://ai.google.dev/docs/function_calling#expandable-5

Idea that user/model and function parts should always be interleaved.
I.e. you can get two parts of the same type after each other. 

And you should call a `function` and provide `functionResponse` straight after model gave you answer. (Although you can ignore `functionCall` and make a `user` content.

API will raise an error if request is not formatted properly.